src/provider/provider_devdax_memory.c

```diff
@@ -228,15 +228,24 @@ static umf_result_t devdax_alloc(void *provider, size_t size, size_t alignment,
         return UMF_RESULT_ERROR_INVALID_ARGUMENT;
     }
 
-    // alignment must be a power of two and a multiple of sizeof(void *)
-    if (alignment &&
-        ((alignment & (alignment - 1)) || (alignment % sizeof(void *)))) {
-        LOG_ERR("wrong alignment: %zu (not a power of 2 or a multiple of "
-                "sizeof(void *))",
-                alignment);
+    // alignment must be a power of two and a multiple or a divider of the page size
+    if (alignment && ((alignment & (alignment - 1)) ||
+                      ((alignment % DEVDAX_PAGE_SIZE_2MB) &&
+                       (DEVDAX_PAGE_SIZE_2MB % alignment)))) {
+        LOG_ERR("wrong alignment: %zu (not a power of 2 or a multiple or a "
+                "divider of the page size (%zu))",
+                alignment, DEVDAX_PAGE_SIZE_2MB);
         return UMF_RESULT_ERROR_INVALID_ARGUMENT;
     }
 
+    if (IS_NOT_ALIGNED(alignment, DEVDAX_PAGE_SIZE_2MB)) {
+        alignment = ALIGN_UP(alignment, DEVDAX_PAGE_SIZE_2MB);
+    }
+
+    if (IS_NOT_ALIGNED(size, DEVDAX_PAGE_SIZE_2MB)) {
+        size = ALIGN_UP(size, DEVDAX_PAGE_SIZE_2MB);
```

Comment:
Do I understand this code correctly that if jemalloc requested 4KB this provider will allocate 2MB and return it?

Reply:
Yes, see also https://github.com/oneapi-src/unified-memory-framework/pull/853 - in case of jemalloc it will be aligned up earlier - in the jemalloc's hook.

Reply:
This approach is aligned to @pbalcer's comment: https://github.com/oneapi-src/unified-memory-framework/pull/838#discussion_r1816917818

Reply:
But will jemalloc actually use the full 2MB or just 4KB?

@pbalcer in his comment mentioned only alignment but you also increase the size. So as I understand if original `size` is `4096` the `ALIGN_UP(size, DEVDAX_PAGE_SIZE_2MB)` will return 2MB. So the pool manager (e.g. jemalloc requested 4096 bytes) the memory provider will allocates 2MB. It is waste of memory because the pool manager does not know that 2MB was allocated and will use only 4096 bytes out of 2MB. 

In general here are my thoughts:

1. The alloc function of any memory provider should not make such assumption and should not allocate more than was requested. 
2. Every pool manager should request the size that is multiple of the min page size of the corresponding provider. It is a contract between Pool Manager and Memory Provider in our API.

Reply:
> But will jemalloc actually use the full 2MB or just 4KB?
> 
Just 4kB.

> @pbalcer in his comment mentioned only alignment but you also increase the size. 
>

No, read carefully: "_I'd put it more strongly, not only alignment, **but also size**, must always be a multiple of a page size._"
https://github.com/oneapi-src/unified-memory-framework/pull/838#discussion_r1816917818

Reply:
I can see only two choices:
1) Do NOT align up a size in `devdax_alloc()` (no waste of memory) and do it later in `devdax_open_IPC_handle()` and `devdax_close_IPC_handle()` OR
2) Align up a size in `devdax_alloc()` (waste of memory) and do NOT that later in `devdax_open_IPC_handle()` and `devdax_close_IPC_handle()` - the current approche.

I cannot see any other solution. Do you?

Reply:
> In general here are my thoughts:
> 
> 1. The alloc function of any memory provider should not make such assumption and should not allocate more than was requested.

... so **we have to do `ALIGN_UP(size, 2MB)`** in `devdax_open_IPC_handle()` and `devdax_close_IPC_handle()` - there is no other choice left.

> 2. Every pool manager should request the size that is multiple of the min page size of the corresponding provider. It is a contract between Pool Manager and Memory Provider in our API.

Exactly that I have just implemented in https://github.com/oneapi-src/unified-memory-framework/pull/853.

Reply:
> I can see only two choices:
> 
> 1. Do NOT align up a size in `devdax_alloc()` (no waste of memory) and do it later in `devdax_open_IPC_handle()` and `devdax_close_IPC_handle()` OR
> 2. Align up a size in `devdax_alloc()` (waste of memory) and do NOT that later in `devdax_open_IPC_handle()` and `devdax_close_IPC_handle()` - the current approche.
> 
> I cannot see any other solution. Can you?

@vinser52 please make a decision which of these two options I should implement. 
Taking into account that you said:

> 1. The alloc function of any memory provider should not make such assumption and should not allocate more than was requested.

I probably have to revert and go back to option no. 1.

Reply:
@vinser52 In `devdax_alloc()` I replaced
```c
    if (IS_NOT_ALIGNED(size, DEVDAX_PAGE_SIZE_2MB)) {
        size = ALIGN_UP(size, DEVDAX_PAGE_SIZE_2MB);
    }
```

with:

```c
    if (IS_NOT_ALIGNED(size, DEVDAX_PAGE_SIZE_2MB)) {
        LOG_ERR("wrong size: %zu (not a multiple of the page size (%zu))", size,
                DEVDAX_PAGE_SIZE_2MB);
        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
    }
```

and as a result the `jemallocCoarseDevDaxTest` tests fail:

```
[  FAILED  ] 5 tests, listed below:
[  FAILED  ] jemallocCoarseDevDaxTest/umfPoolTest.pow2AlignedAlloc/0, where GetParam() = (0x55658d97e220, NULL, 0x7feff6760200, 0x55658d97e770, 0x55658d97e740)
[  FAILED  ] jemallocCoarseDevDaxTest/umfPoolTest.multiThreadedpow2AlignedAlloc/0, where GetParam() = (0x55658d97e220, NULL, 0x7feff6760200, 0x55658d97e770, 0x55658d97e740)
[  FAILED  ] jemallocCoarseDevDaxTest/umfPoolTest.malloc_compliance/0, where GetParam() = (0x55658d97e220, NULL, 0x7feff6760200, 0x55658d97e770, 0x55658d97e740)
[  FAILED  ] jemallocCoarseDevDaxTest/umfPoolTest.calloc_compliance/0, where GetParam() = (0x55658d97e220, NULL, 0x7feff6760200, 0x55658d97e770, 0x55658d97e740)
[  FAILED  ] jemallocCoarseDevDaxTest/umfPoolTest.realloc_compliance/0, where GetParam() = (0x55658d97e220, NULL, 0x7feff6760200, 0x55658d97e770, 0x55658d97e740)

 5 FAILED TESTS
```

because of the errors like:
```
[PID:31753  TID:31753  ERROR UMF] devdax_alloc: wrong size: 2621440 (not a multiple of the page size (2097152))
[PID:31753  TID:31753  ERROR UMF] devdax_alloc: wrong size: 790528 (not a multiple of the page size (2097152))
```

because jemalloc wants to alloc a size not being a multiple of 2MB in `arena_extent_alloc()` called from `mallocx()`, for example:

```
#0  devdax_alloc (provider=0x7ffff7677008, size=2621440, alignment=2097152, resultPtr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/provider/provider_devdax_memory.c:248
#1  0x00007ffff7f7fb83 in umfMemoryProviderAlloc (hProvider=0x7ffff7711168, size=2621440, alignment=4096, ptr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/memory_provider.c:230
#2  0x00007ffff7f85c33 in coarse_memory_provider_alloc (provider=0x7ffff771a0e8, size=2621440, alignment=4096, resultPtr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/provider/provider_coarse.c:1304
#3  0x00007ffff7f7fb83 in umfMemoryProviderAlloc (hProvider=0x7ffff7711268, size=2621440, alignment=4096, ptr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/memory_provider.c:230
#4  0x00007ffff7f8eaf5 in trackingAlloc (hProvider=0x7ffff771f0a8, size=2621440, alignment=4096, ptr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/provider/provider_tracking.c:164
#5  0x00007ffff7f7fb83 in umfMemoryProviderAlloc (hProvider=0x7ffff7711368, size=2621440, alignment=4096, ptr=0x7fffffffb5a8) at /home/ldorau/work/unified-memory-framework/src/memory_provider.c:230
#6  0x0000555555596d11 in arena_extent_alloc (extent_hooks=0x5555556411c0 <arena_extent_hooks>, new_addr=0x0, size=2621440, alignment=4096, zero=0x7fffffffb627, commit=0x7fffffffb626, arena_ind=50)
    at /home/ldorau/work/unified-memory-framework/src/pool/pool_jemalloc.c:79
#7  0x00007ffff7d01dad in ?? () from /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
#8  0x00007ffff7d02b25 in ?? () from /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
#9  0x00007ffff7d0fa53 in ?? () from /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
#10 0x00007ffff7cbbe30 in mallocx () from /usr/lib/x86_64-linux-gnu/libjemalloc.so.2
```

The jemalloc's page size (`arenas.page` - see https://jemalloc.net/jemalloc.3.html) is read-only and cannot be changed from the default 4096 to 2MB, so we cannot require the size in `devdax_alloc()` to be a multiple of 2MB.

Reply:
@vinser52 See the build: https://github.com/oneapi-src/unified-memory-framework/actions/runs/11610652104/job/32330370629
of https://github.com/oneapi-src/unified-memory-framework/pull/855

Reply:
@vinser52 Done.